batch.go

```diff
@@ -0,0 +1,197 @@
+	buf = snappy.Encode(nil, buf)
+	return buf, entriesCount, nil
+}
+
+// encodeJSON encodes the batch as JSON push request, and returns the encoded
+// bytes and the number of encoded entries
+func (b *Batch) encodeJSON() ([]byte, int, error) {
+	req, entriesCount := b.createPushRequest()
+	buf, err := json.Marshal(req)
+	if err != nil {
+		return nil, 0, err
+	}
+	return buf, entriesCount, nil
+}
+
+// createPushRequest creates a push request and returns it, together with
+// number of entries
+func (b *Batch) createPushRequest() (*logproto.PushRequest, int) {
+	req := logproto.PushRequest{
+		Streams: make([]logproto.Stream, 0, len(b.Streams)),
+	}
+
+	entriesCount := 0
+	for _, stream := range b.Streams {
+		req.Streams = append(req.Streams, *stream)
+		entriesCount += len(stream.Entries)
+	}
+	return &req, entriesCount
+}
+
+// generateEntries creates a list of log entries
+func generateEntries(ctx context.Context, tenantID string, pool LabelPool, s int, min int, max int) []Entry {
+	state := lib.GetState(ctx)
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "localhost"
+	}
+
+	result := make([]Entry, 0)
+	var labels model.LabelSet
+	var n int
+	// create `[min,max)` entries for `s` streams
+	for i := 0; i < s; i++ {
+		labels = labelsFromPool(pool)
+		labels[model.InstanceLabel] = model.LabelValue(fmt.Sprintf("vu%d.%s.local", state.VUID, hostname))
+
+		n = min + rand.Intn(max-min)
+		entries := make([]Entry, n)
+		for i := 0; i < n; i++ {
+			entries[i] = Entry{
+				Entry: logproto.Entry{
+					Timestamp: time.Now(),
+					Line:      flog.NewLog(string(labels[model.LabelName("format")]), time.Now()),
+				},
+				TenantID: tenantID,
+				Labels:   labels,
+			}
+			time.Sleep(10 * time.Millisecond)
+		}
+		result = append(result, entries...)
+	}
+	return result
+}
+
+// choice returns a single label value from a list of label values
+func choice(values model.LabelValues) model.LabelValue {
+	return values[rand.Intn(values.Len())]
+}
+
+// labelsFromPool creates a label set from the given label value pool `p`
+func labelsFromPool(p LabelPool) model.LabelSet {
+	return model.LabelSet{
+		"app":       choice(p.App),
+		"format":    choice(p.Format),
+		"namespace": choice(p.Namespace),
+		"pod":       choice(p.Pod),
+	}
+}
+
+// generateValues returns `n` label values generated with the `ff` gofakeit function
+func generateValues(ff FakeFunc, n int) model.LabelValues {
+	res := make(model.LabelValues, n)
+	for i := 0; i < n; i++ {
+		res[i] = model.LabelValue(ff())
+	}
+	return res
+}
+
+// newLabelPool creates a "pool" of values for each label name
+// The amount of different values per label name is not configurable yet.
+func newLabelPool(faker *fake.Faker) LabelPool {
+	return LabelPool{
+		Format:    model.LabelValues{"apache_common", "json"}, // needs to match the available flog formats
+		Namespace: generateValues(faker.BS, 10),
+		App:       generateValues(faker.BS, 5),
+		Pod:       generateValues(faker.BS, 100),
```

Comment:
Ideally, there would be JSON definition of these labels so they are not static.

Reply:
You can take a look at https://github.com/grafana/xk6-client-prometheus-remote/blob/a5db579b304270ae80ca0f689fe35b77c2156041/remote_write.go#L342-L348 if that is what you mean?

Reply:
Yeah now that I think about if we generated those randomly, querying will be hard. So we might want to go for namespace1,namespace2.....

Or go with something like cortex.